### PR TITLE
Checks for K-Means (batch and distributed)

### PR DESCRIPTION
**Pull request by amgrigoriev**

### CODE REVIEW

**Review by KulikovNikita — APPROVED** [review]

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i [hunk 1]
```diff
@@ -122,43 +130,30 @@ Status KMeansDistributedStep1KernelUCAPI<algorithmFPType>::compute(size_t na, co
...
-                DAAL_CHECK_STATUS_VAR(st);
+                DAAL_CHECK_STATUS_VAR(this->computePartialCandidates(assignments, range.count, nClusters, int(block == 0)));
+                DAAL_CHECK_STATUS_VAR(this->mergePartialCandidates(nClusters));
             }
             needCandidates = hasEmptyClusters;
         }
-        this->updateObjectiveFunction(outObjFunction, range.count, nClusters, int(block == 0), st);
-        DAAL_CHECK_STATUS_VAR(st);
+        DAAL_CHECK_STATUS_VAR(this->updateObjectiveFunction(outObjFunction, range.count, nClusters, int(block == 0)));
         ntData->releaseBlockOfRows(dataRows);
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i:157]: Check status for releaseBlockOfRows.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i [hunk 2]
```diff
@@ -170,8 +165,7 @@ Status KMeansDistributedStep1KernelUCAPI<algorithmFPType>::compute(size_t na, co
             DAAL_CHECK_STATUS_VAR(ntAssignments->releaseBlockOfRows(assignmentsRows));
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i:165]: Check buffer before context.copy call.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i [hunk 3]
```diff
@@ -170,8 +165,7 @@ Status KMeansDistributedStep1KernelUCAPI<algorithmFPType>::compute(size_t na, co
...
         }
     }
-    this->mergeReduceCentroids(outCentroids, nClusters, nFeatures, st);
-    DAAL_CHECK_STATUS_VAR(st);
+    DAAL_CHECK_STATUS_VAR(this->mergeReduceCentroids(outCentroids, nClusters, nFeatures));
     context.copy(outCCounters, 0, this->_partialCentroidsCounters, 0, nClusters, st);
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i:169]: Check buffer before context.copy call.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i [hunk 4]
```diff
@@ -170,8 +165,7 @@ Status KMeansDistributedStep1KernelUCAPI<algorithmFPType>::compute(size_t na, co
...
     DAAL_CHECK_STATUS_VAR(st);
     DAAL_CHECK_STATUS_VAR(ntInCentroids->releaseBlockOfRows(inCentroidsRows));
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i:171]: Check buffer before context.copy call few lines below.

Comment (vepifanov) [reply cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step1_ucapi_impl.i:171]: Check buffer before toHost call few lines below.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i [hunk 5]
```diff
@@ -91,11 +100,11 @@ Status KMeansDistributedStep2KernelUCAPI<algorithmFPType>::compute(size_t na, co
...
-    const size_t nBlocks = na / 5;
+    const uint32_t nBlocks = static_cast<uint32_t>(na) / 5;
 
     algorithmFPType tmpObjValue = 0.0;
 
-    for (size_t i = 0; i < nBlocks; i++)
+    for (uint32_t i = 0; i < nBlocks; i++)
     {
         BlockDescriptor<int> ntParClusterS0Rows;
         const_cast<NumericTable *>(a[i * 5 + 0])->getBlockOfRows(0, nClusters, readOnly, ntParClusterS0Rows);
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i:110]: Check status for getBlockOfRows.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i [hunk 6]
```diff
@@ -105,13 +114,14 @@ Status KMeansDistributedStep2KernelUCAPI<algorithmFPType>::compute(size_t na, co
         const_cast<NumericTable *>(a[i * 5 + 1])->getBlockOfRows(0, nClusters, readOnly, ntParClusterS1Rows);
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i:114]: Check status for getBlockOfRows.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i [hunk 7]
```diff
@@ -105,13 +114,14 @@ Status KMeansDistributedStep2KernelUCAPI<algorithmFPType>::compute(size_t na, co
...
         auto inParClusterS1 = ntParClusterS1Rows.getBuffer();
 
-        updateClusters(i == 0, inParClusterS0, inParClusterS1, outCCounters, outCentroids, nClusters, nFeatures, st);
+        DAAL_CHECK_STATUS_VAR(updateClusters(i == 0, inParClusterS0, inParClusterS1, outCCounters, outCentroids, nClusters, nFeatures));
 
         BlockDescriptor<algorithmFPType> ntParObjFunctionRows;
         const_cast<NumericTable *>(a[i * 5 + 2])->getBlockOfRows(0, 1, readOnly, ntParObjFunctionRows);
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i:120]: Check status for getBlockOfRows.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i [hunk 8]
```diff
@@ -105,13 +114,14 @@ Status KMeansDistributedStep2KernelUCAPI<algorithmFPType>::compute(size_t na, co
...
+        DAAL_ASSERT(ntParObjFunctionRows.getBuffer().size() > 0);
         auto inParObjFunction = ntParObjFunctionRows.getBuffer().toHost(data_management::readOnly, st);
         DAAL_CHECK_STATUS_VAR(st);
-        tmpObjValue += inParObjFunction.get()[0];
+        tmpObjValue += *inParObjFunction.get();
 
         BlockDescriptor<algorithmFPType> ntParCValuesRows;
         const_cast<NumericTable *>(a[i * 5 + 3])->getBlockOfRows(0, nClusters, readOnly, ntParCValuesRows);
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i:127]: Check status for getBlockOfRows.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i [hunk 9]
```diff
@@ -121,17 +131,20 @@ Status KMeansDistributedStep2KernelUCAPI<algorithmFPType>::compute(size_t na, co
         const_cast<NumericTable *>(a[i * 5 + 4])->getBlockOfRows(0, nClusters, readOnly, ntParCCandidates);
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i:131]: Check status for getBlockOfRows.

File: cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i [hunk 10]
```diff
@@ -121,17 +131,20 @@ Status KMeansDistributedStep2KernelUCAPI<algorithmFPType>::compute(size_t na, co
...
     }
-    outObjFunction.get()[0] = tmpObjValue;
+    *outObjFunction.get() = tmpObjValue;
     {
+        DAAL_ASSERT(outCentroids.size() >= nFeatures * nClusters);
         auto retCentroids = outCentroids.toHost(ReadWriteMode::readWrite, st);
+        DAAL_CHECK_STATUS_VAR(st);
+        DAAL_ASSERT(outCentroids.size() >= nClusters);
         auto retCCounters = outCCounters.toHost(ReadWriteMode::readOnly, st);
         DAAL_CHECK_STATUS_VAR(st);
```
Comment (vepifanov) [inline cpp/daal/src/algorithms/kmeans/oneapi/kmeans_lloyd_distr_step2_ucapi_impl.i:150]: Check status for releaseBlockOfRows few lines below.

**Review by vepifanov — COMMENTED** [review]
Add buffer checks before context.copy/context.fill calls.

**Review by vepifanov — APPROVED** [review]

**Pull request merged**